### PR TITLE
iOS: Added local notification in terminate function.

### PR DESCRIPTION
**Pull request by eyganeshv**
## Description
#110 

Add Local Push Notification for OnTerminate() - Android #110
Fixed this issue for iOS

## How to test
Notification will appear when app killed.

### CODE REVIEW

File: ios/COVIDSafePaths/AppDelegate.m [hunk 1]
```diff
@@ -80,6 +80,16 @@ - (void)application:(UIApplication *)application didReceiveLocalNotification:(UI
 }
 
 - (void)applicationWillTerminate:(UIApplication *)application {
+    UILocalNotification *_localNotification = [[UILocalNotification alloc] init];
+    _localNotification.fireDate = [NSDate dateWithTimeIntervalSinceNow:5];
+    _localNotification.timeZone = [NSTimeZone defaultTimeZone];
+    _localNotification.alertTitle = @"Private Kit Enabled";
```
Comment (tstirrat) [inline ios/COVIDSafePaths/AppDelegate.m:86]: Use the app name: COVID Safe Paths

File: ios/COVIDSafePaths/AppDelegate.m [hunk 2]
```diff
@@ -80,6 +80,16 @@ - (void)application:(UIApplication *)application didReceiveLocalNotification:(UI
...
+    _localNotification.alertBody = @"Private Kit is securely storing your GPS coordinates once every five minutes on this device.";
```
Comment (tstirrat) [inline ios/COVIDSafePaths/AppDelegate.m:87]: Use the app name: COVID Safe Paths is securely...

File: ios/COVIDSafePaths/AppDelegate.m (hunk 1 above)
Comment (eyganeshv) [reply ios/COVIDSafePaths/AppDelegate.m:86]: Sure I will update

File: ios/COVIDSafePaths/AppDelegate.m (hunk 2 above)
Comment (eyganeshv) [reply ios/COVIDSafePaths/AppDelegate.m:87]: Sure I will update

**Comment by tstirrat** [comment]
This happens every time the app is closing. I think it will get annoying if someone is checking each day, or multiple times per day.

If you could make it show the warning only the first time they close the app, that would be useful.

@kenpugsley @penrods thoughts?

**Comment by eyganeshv** [comment]
I can update show the warning at first time when close the app only. Please confirm?

**Comment by tremblerz** [comment]
Yes, I think the first time only warning sounds better.

File: ios/COVIDSafePaths/AppDelegate.m [hunk 3]
```diff
@@ -80,6 +81,27 @@ - (void)application:(UIApplication *)application didReceiveLocalNotification:(UI
...
+        _localNotification.alertBody = @"COVID Safe Paths is securely storing your GPS coordinates once every five minutes on this device.";
+        _localNotification.soundName = UILocalNotificationDefaultSoundName;
+        [[UIApplication sharedApplication]scheduleLocalNotification:_localNotification];
+      }
+    }
+    [[NSUserDefaults standardUserDefaults] setBool:YES forKey:@"NotificationAtFirstTimeOnly"];
+    [[NSUserDefaults standardUserDefaults] synchronize];
+  }
+    
+  //  ***********************************************************************************//
```
Comment (tstirrat) [inline ios/COVIDSafePaths/AppDelegate.m:104]: is the code below doing a similar thing? can we remove it? @tremblerz

Comment (eyganeshv) [reply ios/COVIDSafePaths/AppDelegate.m:104]: Mean, to remove from line number 105 to 113?

Comment (tstirrat) [reply ios/COVIDSafePaths/AppDelegate.m:104]: Yes. I think so.


When does that notification show?

Comment (eyganeshv) [reply ios/COVIDSafePaths/AppDelegate.m:104]: @tremblerz Plz confirm. Shall I remove from line number 105 to 113?

Comment (eyganeshv) [reply ios/COVIDSafePaths/AppDelegate.m:104]: @tstirrat I think that notification show when app quit [Code line number 105 to 113]. It is not working in my iPhone.

Comment (tstirrat) [reply ios/COVIDSafePaths/AppDelegate.m:104]: Yes, please remove the other notification, it's trying to do the same thing.

File: ios/COVIDSafePaths/AppDelegate.m [hunk 4]
```diff
@@ -80,15 +81,25 @@ - (void)application:(UIApplication *)application didReceiveLocalNotification:(UI
...
+  if(![[NSUserDefaults standardUserDefaults] boolForKey:@"NotificationAtFirstTimeOnly"]) {
+    if ([[UIApplication sharedApplication] respondsToSelector:@selector(currentUserNotificationSettings)]){
+      //Checking local notification permission or not.
+      UIUserNotificationSettings *grantedSettings = [[UIApplication sharedApplication] currentUserNotificationSettings];
+      if (grantedSettings.types != UIUserNotificationTypeNone){
+        // Set local notification.
+        UILocalNotification *_localNotification = [[UILocalNotification alloc] init];
+        _localNotification.fireDate = [NSDate dateWithTimeIntervalSinceNow:5];
+        _localNotification.timeZone = [NSTimeZone defaultTimeZone];
+        _localNotification.alertTitle = @"COVID Safe Paths";
```
Comment (tstirrat) [inline ios/COVIDSafePaths/AppDelegate.m:94]: Thanks, and now can you sync with `develop` branch to add these strings into en.lproj and localize them?

e.g.

https://github.com/tripleblindmarket/covid-safe-paths/blob/5f2f131b57018e7ee849e829a028a3e78b301755/ios/en.lproj/Localizable.strings#L2-L5

https://github.com/tripleblindmarket/covid-safe-paths/blob/5f2f131b57018e7ee849e829a028a3e78b301755/ios/COVIDSafePaths/AppDelegate.m#L83-L84

Comment (eyganeshv) [reply ios/COVIDSafePaths/AppDelegate.m:94]: Ok. I will update.

Comment (eyganeshv) [reply ios/COVIDSafePaths/AppDelegate.m]: @tstirrat Merged my branch developer into ganesh-ios. You can check two file AppDelegate.m and  en.lproj/Localizable.strings only.

File: ios/COVIDSafePaths/COVIDSafePaths.entitlements [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (tstirrat) [inline ios/COVIDSafePaths/COVIDSafePaths.entitlements:1]: not sure what this file is . remove

File: ios/COVIDSafePaths/AppDelegate.m [hunk 6]
```diff
@@ -78,15 +79,25 @@ - (void)application:(UIApplication *)application didReceiveLocalNotification:(UI
...
-
-  [RNCPushNotificationIOS didReceiveLocalNotification:notification];
+  //Show local notifiation at first time only.
+  if(![[NSUserDefaults standardUserDefaults] boolForKey:@"NotificationAtFirstTimeOnly"]) {
+    if ([[UIApplication sharedApplication] respondsToSelector:@selector(currentUserNotificationSettings)]){
+      //Checking local notification permission or not.
+      UIUserNotificationSettings *grantedSettings = [[UIApplication sharedApplication] currentUserNotificationSettings];
+      if (grantedSettings.types != UIUserNotificationTypeNone){
+        // Set local notification.
+        UILocalNotification *_localNotification = [[UILocalNotification alloc] init];
```
Comment (tstirrat) [inline ios/COVIDSafePaths/AppDelegate.m:89]: Great, looking good. 

This whole method is getting large and there are mixed levels of abstraction. You've added good comments, yet I think those would be better as separate methods:

(pseudo code)

```js
applicationWillTerminate() {
  if (isFirstTimeClosing() && hasNotificationPermissions()) {
    notifThatWeAreStillTracking();  // <--- if this is moved to a new method the applicationWillTerminate method is easier to understand
  }
}
```

File: ios/COVIDSafePaths/COVIDSafePaths.entitlements (hunk 5 above)
Comment (eyganeshv) [reply ios/COVIDSafePaths/COVIDSafePaths.entitlements:1]: No use this. Will remove.

File: ios/COVIDSafePaths/AppDelegate.m (hunk 6 above)
Comment (eyganeshv) [reply ios/COVIDSafePaths/AppDelegate.m:89]: I will do.

**Pull request merged**